File: output/deepseek_r1/avatar/algo_based_translation/python/java/successful_translation/atcoder_ABC122_D.java

import java.util.*;

public class atcoder_ABC122_D {
    static int n;
    static final int mod = (int)1e9 + 7;
    static Map<String, Integer>[] memo;

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        n = scanner.nextInt();
        memo = new HashMap[n + 1];
        for (int i = 0; i <= n; i++) {
            memo[i] = new HashMap<>();
        }
        System.out.println(dfs(0, "TTT"));
    }

    static boolean ok(String last4) {
        for (int i = 0; i < 4; i++) {
            char[] arr = last4.toCharArray();
            if (i >= 1) {
                char temp = arr[i - 1];
                arr[i - 1] = arr[i];
                arr[i] = temp;
            }
            if (new String(arr).contains("AGC")) {
                return false;
            }
        }
        return true;
    }

    static int dfs(int cur, String last3) {
        if (memo[cur].containsKey(last3)) {
            return memo[cur].get(last3);
        }
        if (cur == n) {
            return 1;
        }
        int ret = 0;
        String chars = "ACGT";
        for (int i = 0; i < chars.length(); i++) {
            char c = chars.charAt(i);
            String s = last3 + c;
            if (ok(s)) {
                String newLast3 = s.substring(1);
                ret = (ret + dfs(cur + 1, newLast3)) % mod;
            }
        }
        memo[cur].put(last3, ret);
        return ret;
    }
}

File: output/llama_4_maverick/avatar/algo_based_translation/java/python/atcoder_ABC122_D.py

MOD = 1000000007

def isOK(last4):
    if "AGC" in last4:
        return False
    for i in range(3):
        vals = list(last4)
        vals[i], vals[i + 1] = vals[i + 1], vals[i]
        s = "".join(vals)
        if "AGC" in s:
            return False
    return True

def dfs(current, last3, N, TOKENS, memo):
    if (current, last3) in memo:
        return memo[(current, last3)]
    if current == N:
        return 1
    result = 0
    for c in TOKENS:
        if isOK(last3 + c):
            result = (result + dfs(current + 1, last3[1:] + c, N, TOKENS, memo)) % MOD
    memo[(current, last3)] = result
    return result

def calc(N):
    TOKENS = ["A", "C", "G", "T"]
    memo = {}
    return dfs(0, "TTT", N, TOKENS, memo)

def main():
    N = int(input())
    print(calc(N))

if __name__ == "__main__":
    main()

